Allow capturing threads without writing to disk

fractalwrench opened the pull request:
## Goal

Captures the thread information from KSCrash without writing to disk. This is required for handled reports, where it's necessary to display the thread information up front so a user can invoke a callback.

This changeset fully populates the `event.threads` and `event.errors` fields before writing to disk.

## Changeset

### Thread capture
- Added a method to `BSG_KSCrash` for deserializing JSON from a char array, and for constructing threads out of a JSON dictionary
- Added `bsg_kscrash_captureThreadTrace()`, which is an internal API for suspending threads and captures their information as a JSON string
- Added convenience methods for suspending/resuming threads
- `BSG_KSCrashReport` has its JSON serializers for thread/stacktrace information to separate methods, which are invoked both when collecting thread information and when writing a report to disk
- Different `BSG_KSCrashReportWriter` implementations are passed when collecting thread information - one writes JSON to a char array, the other writes JSON to disk
- The JSON codec does not requires a null terminator when writing to disk, this is added when copying to the char array by using a temporary variable

### Stacktrace collection
- Updated default discard depth to 3 and made all `notify` calls within `BugsnagClient` have a consistent count of stackframes by inlining implementations where appropriate
- Removed unused parameters from `reportUserException` on `BugsnagCrashSentry`
- Fill out context fields in `bsg_kscrash_captureThreadTrace` which are required for stacktrace serialization
- Extracted stacktrace collection from `bsg_kscrashsentry_reportUserException` and `BugsnagClient` into `captureThreads`

## Tests

Added a unit test for JSON deserialization and verified that the thread information is displayed in the dashboard for an example app.

## TBD

- address E2E test failures (e.g. swift fatalError)
- remove unused parameters?

tomlongridge commented on Source/KSCrash/Source/KSCrash/Recording/Sentry/BSG_KSCrashSentry_User.c:
```diff
@@ -147,10 +143,19 @@ void bsg_kscrashsentry_reportUserException(const char *name, const char *reason,
+    bsg_kscrashsentry_suspendThreads();
+}
+
+void bsg_kscrashsentry_resume_threads_user() {
```
Worth adding a `terminateProgram` parameter so that it can be called from `bsg_kscrashsentry_reportUserException` too?

fractalwrench replied:
Added parameter and removed duplicate code.

tomlongridge commented on Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrashReport.c:
```diff
@@ -1635,3 +1617,59 @@ void bsg_kscrashreport_logCrash(const BSG_KSCrash_Context *const crashContext) {
+}
+
+char *bsg_kscrw_i_captureThreadTrace(const BSG_KSCrash_Context *crashContext) {
+    if (thread_json_data == NULL) {
+        thread_json_data = malloc(thread_json_size);
+    }
+    memset(thread_json_data, 0, strlen(thread_json_data));
+    thread_json_data[0] = '\0';
```
Can you move this to `bsg_kscrw_i_collectJsonData` so `thread_json_data` is managed in one place?

fractalwrench replied:
Extracted to `bsg_kscrw_i_resetThreadTraceData` as suggested by in-person discussion.

tomlongridge commented on Source/KSCrash/Source/KSCrash/Recording/Sentry/BSG_KSCrashSentry_User.c:
```diff
@@ -101,11 +101,7 @@ void bsg_kscrashsentry_reportUserException(const char *name, const char *reason,
         int callstackCount = 100;
         uintptr_t callstack[callstackCount];
 
-        if (localContext->suspendThreadsForUserReported) {
```
Why are we not honouring `suspendThreadsForUserReported`?

fractalwrench replied:
This is always set to `true` and is an internal property of KSCrash which isn't exposed. For now I've reinstated the checks here but it would make sense to fully remove `suspendThreadsForUserReported` at a later date as it's not used.

fractalwrench commented on Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrashReport.c:
```diff
@@ -1635,3 +1617,59 @@ void bsg_kscrashreport_logCrash(const BSG_KSCrash_Context *const crashContext) {
+
+int bsg_kscrw_i_collectJsonData(const char *const data, const size_t length, void *const userData) {
+    size_t new_size = strlen(thread_json_data) + length;
+    char tmp[length];
+    memcpy(tmp, data, length);
+    tmp[length] = '\0';
```
KSCrash does not pass non-terminated chars by default in its JSON codec. Altering the codec itself to use terminated chars resulted in KSCrash being unable to read the serialized report - but the `NSJSONSerialization` class does seem to require null-terminated char arrays.

fractalwrench replied:
Action: increase length of `tmp` by 1 to avoid overflow, and rename to `terminatedString`

fractalwrench commented:
Updated to fix the serialization of error reporting thread's stacktrace, which is recorded differently compared to all the other threads.

I've converted this to a draft PR in the meanwhile as there are a few more things that need testing out. Once that's done, the changeset should fix the population of both `event.threads` and `event.errors`.

fractalwrench commented on Source/BugsnagClient.m:
```diff
@@ -935,6 +940,7 @@ - (void)notifyInternal:(BugsnagEvent *_Nonnull)event
                  block:(BugsnagOnErrorBlock)block
 {
     [self populateEventData:event];
+    event.threads = [self generateThreads];
```
This is now called from within `notify:handledState:block:` which is where every other field is now poulated.

fractalwrench commented on Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrashReport.c:
```diff
@@ -1635,3 +1617,59 @@ void bsg_kscrashreport_logCrash(const BSG_KSCrash_Context *const crashContext) {
+    {
+    // Conditionally write threads depending on user configuration
+    int sendPolicy = crashContext->crash.threadTracingEnabled;
+    bool unhandledCrash = crashContext->crash.crashType != BSG_KSCrashTypeUserReported;
```
Moved to top of function.

fractalwrench commented on Source/BugsnagClient.m:
```diff
@@ -1475,6 +1481,44 @@ - (BugsnagDeviceWithState *)generateDeviceWithState:(NSDictionary *)systemInfo {
+        }
+    }
+    return nil;
+}
```
Moved function to `BSGSerialization`.

fractalwrench commented on Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrashC.h:
```diff
@@ -182,6 +182,14 @@ void bsg_kscrash_setWriteBinaryImagesForUserReported(
 BSG_KSCrash_Context *crashContext(void);
 
+/**
+ * Captures a thread trace for the current application state, if the user
+ * has configured this functionality.
```
`bsg_kscrw_i_writeAllThreads` takes a parameter `recordAllThreads` which controls whether all threads are captured. If this parameter is false, the threads are still suspended but only the main thread's stacktrace is serialized.

I've updated the function's documentation accordingly

tomlongridge commented on Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrash.m:
```diff
@@ -306,41 +316,64 @@ - (void)deleteAllReports {
+        callstack[i] = [addresses[i] unsignedLongValue];
+    }
+    if (numFrames > 0) {
+        depth = 0; // reset depth if the stack does not need to be generated
```
This took some parsing in my head!

I think it would make the intent clearer if the work to populate callstack from the exception happens inside the `numFrames > 0` path, then we don't need to realloc and some comments can describe the two cases separately.

fractalwrench replied:
Refactored this so that callstack allocation happens within the `numFrames > 0` path.

tomlongridge commented on Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrashReport.c:
```diff
@@ -1635,3 +1619,61 @@ void bsg_kscrashreport_logCrash(const BSG_KSCrash_Context *const crashContext) {
+    memcpy(terminated_str, data, length);
+    terminated_str[length] = '\0';
+
+    if (new_size > bsg_g_thread_json_size) {
```
I believe the data may or may not have a terminator already (I can see calls from KS code that passes data as a string literal which will be terminated.

Therefore by adding our own and then `strlen`-ing the array, we will be able to pass it to `strlen` to give us the length we need. So I think you can do away with `new_size` and just use:

```suggestion
    if (strlen(bsg_g_thread_json_data) + strlen(terminated_str) > bsg_g_thread_json_size) {
```

fractalwrench replied:
Updated code to use this suggestion.

tomlongridge commented on Source/KSCrash/Source/KSCrash/Recording/Sentry/BSG_KSCrashSentry_User.c:
```diff
@@ -96,34 +93,12 @@ void bsg_kscrashsentry_reportUserException(const char *name, const char *reason,
 
-        // We want these variables to persist until the onCrash
-        // call later
-        int callstackCount = 100;
```
I'm unclear what this function is now doing now - does it need to suspend threads?

tomlongridge commented on Source/KSCrash/Source/KSCrash/Recording/Sentry/BSG_KSCrashSentry_User.h:
```diff
@@ -54,32 +54,29 @@ void bsg_kscrashsentry_uninstallUserExceptionHandler(void);
-        int discardDepth,
         bool terminateProgram);
 
+void bsg_kscrashsentry_suspend_threads_user(void);
+
+void bsg_kscrashsentry_resume_threads_user(bool terminateProgram);
```
Doc comments please

fractalwrench replied:
Added docs to the header file.

fractalwrench commented on Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrash.m:
```diff
@@ -306,41 +316,64 @@ - (void)deleteAllReports {
 
+- (NSArray<BugsnagThread *> *)captureThreads:(NSException *)exc depth:(int)depth {
+    NSArray *addresses = [exc callStackReturnAddresses];
+    int numFrames = (int) [addresses count];
```
`[addresses count]` returns an `NSUInteger` which is an unsigned long - this requires casting to an `int` as this is what KSCrash uses for stackframe counts throughout.

tomlongridge replied:
It didn't before, which is why I was questioning it. Why is this different?

fractalwrench replied:
Previously KSCrash stored the value of `backtrace()` in an `int` as it was calculated in a different file. `backtrace()` returns an `int` which is why this has been changed here:
https://github.com/bugsnag/bugsnag-cocoa/pull/601/files#diff-14847c7e626ff843aeb8c965b155b829L115

fractalwrench commented on Source/BugsnagClient.m:
```diff
@@ -976,36 +981,51 @@ - (void)notifyInternal:(BugsnagEvent *_Nonnull)event
+        }
+    }
+
+    BugsnagError *error = [[BugsnagError alloc] initWithEvent:@{} errorReportingThread:errorReportingThread];
```
Created `initWithErrorReportingThread` as a wrapper around this constructor to better signify the event information is not required for this method of instantiation.

fractalwrench commented on Source/BugsnagClient.m:
```diff
@@ -929,6 +941,21 @@ - (void)notify:(NSException *)exception
+     */
+    int depth = (int)(BSGNotifierStackFrameCount + event.depth);
+    event.threads = [[BSG_KSCrash sharedInstance] captureThreads:event.originalError depth:depth];
+    event.errors = [self generateErrors:event];
```
Refactored so that the errors/threads are captured before creating a `BugsnagEvent` object, avoiding unnecessary duplication.

fractalwrench commented on Source/BugsnagClient.m:
```diff
@@ -822,7 +828,13 @@ - (NSString *)context {
 
 - (void)notifyError:(NSError *_Nonnull)error {
-    [self notifyError:error block:nil];
+    BugsnagHandledState *state = [BugsnagHandledState handledStateWithSeverityReason:HandledError
+                                                                            severity:BSGSeverityWarning
+                                                                           attrValue:error.domain];
+    NSException *wrapper = [NSException exceptionWithName:NSStringFromClass([error class])
+                                                   reason:error.localizedDescription
+                                                 userInfo:error.userInfo];
+    [self notify:wrapper handledState:state block:nil];
```
I've reverted this change as I don't believe it's necessary

fractalwrench commented on Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrash.m:
```diff
@@ -306,41 +316,66 @@ - (void)deleteAllReports {
+        }
+    }
+
+    char *trace = bsg_kscrash_captureThreadTrace(depth, numFrames, callstack);
```
This method currently returns a global variable which can be mutated on another thread once `bsg_kscrash_captureThreadTrace` has resumed thread execution.

tomlongridge commented on Source/BugsnagClient.m:
```diff
@@ -981,36 +1001,55 @@ - (void)notifyInternal:(BugsnagEvent *_Nonnull)event
     [self flushPendingReports];
 }
 
+- (BugsnagError *)generateError:(NSException *)exception
+                                 threads:(NSArray<BugsnagThread *> *)threads {
+    NSString *errorClass = exception.name ?: NSStringFromClass([exception class]);
+    NSString *errorMessage = exception.reason;
+
+    BugsnagThread *errorReportingThread;
+
+    for (BugsnagThread *thread in threads) {
+        if (thread.errorReportingThread) {
+            errorReportingThread = thread;
+            break;
+        }
+    }
+
+    BugsnagError *error = [[BugsnagError alloc] initWithErrorReportingThread:errorReportingThread];
+    error.errorClass = errorClass;
+    error.errorMessage = errorMessage ?: @"";
+    return error;
+}
```
I don't see any reason we can't use `initWithErrorClass` for this - we basically just want to set class, message and type. Passing in the `errorReportingThread` value doesn't seem to do anything with an empty `event` parameter.

fractalwrench replied:
Updated to use existing constructor.

tomlongridge commented on Source/BugsnagEvent.m:
```diff
@@ -569,6 +569,11 @@ - (instancetype)initWithUserData:(NSDictionary *)event {
+
+    if ([errors count] > 0) {
+        BugsnagError *bugsnagError = errors[0];
+        obj.customException = BSGParseCustomException(event, bugsnagError.errorClass, bugsnagError.errorMessage);
```
Should the two strings be copied like in `initWithKSCrashData`? Also, can you rename `bugsnagError` to something else as this implies to me that it's some sort of internally generated construct.

fractalwrench replied:
I don't see any reason why `initWithKSCrashData` needs to copy the string - I assume this is just done for legacy reasons. Renamed variable.

tomlongridge commented on Source/KSCrash/Source/KSCrash/Recording/BSG_KSCrashReport.c:
```diff
@@ -1635,3 +1621,65 @@ void bsg_kscrashreport_logCrash(const BSG_KSCrash_Context *const crashContext) {
+int bsg_kscrw_i_collectJsonData(const char *const data, const size_t length, void *const userData) {
+    if (bsg_kscrw_i_exceedsBufferLen(length)) {
+        bsg_g_thread_json_size += bsg_g_buffer_increment;
+        bsg_g_thread_json_data = realloc(bsg_g_thread_json_data, bsg_g_thread_json_size);
```
It seems to be an anti-pattern to assign the result straight back to the original pointer as it doesn't all you to free the memory. I guess in this case it's a single static area, but would it not be cleaner to `free` and return the error code immediately after `realloc` if it fails rather than introduce a second if-clause to check this later?

fractalwrench replied:
Implemented suggestion and returned immediately in the error case.

fractalwrench commented on Source/BugsnagClient.m:
```diff
@@ -1502,9 +1541,9 @@ - (NSArray *)collectBreadcrumbs {
 - (NSArray *)collectThreads {
-    return @[
-        // TODO implement
-    ];
+    int depth = (int)(BSGNotifierStackFrameCount);
+    NSException *exc = [NSException exceptionWithName:@"Bugsnag" reason:@"" userInfo:nil];
+    return [[BSG_KSCrash sharedInstance] captureThreads:exc depth:depth];
 }
```
This will be used by React Native.

tomlongridge replied:
Can you some comment docs on this, especially as it's a bit of a quirky thing to do!
